language_load.py

```diff
@@ -1,99 +1,152 @@
-        continue
-
-    if not data['id'] in strings_by_object:
-        if verbose:
-            print("No translations for " + data['id'] + " in dump file -- skipping")
-        continue
+    if not translatable(verbose, data):
+        return
+    if not is_object_translated(verbose, obj_id, strings_by_object):
+        return
 
     updated = False
     for string_key in data['strings']:
 
-        if not string_key in strings_by_object[data['id']]:
-            if verbose:
-                print("No translation for " + data['id'] + " string '" + string_key + "' in dump file -- skipping")
+        if not is_key_translated(verbose, obj_id, string_key, strings_by_object[obj_id]):
             continue
 
-        if not fallback_language in data['strings'][string_key]:
-            if verbose:
-                print("No en-GB reference for " + data['id'] + " string '" + string_key + "' in dump file -- probably shouldn't exist; skipping")
+        if not fallback_key_exists(verbose, ref_lang, string_key, obj_id,
+                                   data['strings'][string_key]):
             continue
 
-        if not language_to_import in data['strings'][string_key]:
-            if strings_by_object[data['id']][string_key] == data['strings'][string_key][fallback_language]:
-                # TODO: Is this desirable behaviour?
-                if verbose:
-                    print("Translation for " + data['id'] + " string '" + string_key + "' is identical to en-GB -- skipping")
-                continue
-
-            if verbose:
-                print("Adding " + data['id'] + " string '" + string_key + "'")
-            data['strings'][string_key][language_to_import] = strings_by_object[data['id']][string_key]
-            updated = True
-        else:
-            if strings_by_object[data['id']][string_key] == data['strings'][string_key][language_to_import]:
-                if verbose:
-                    print("Translation for " + data['id'] + " string '" + string_key + "' has not changed -- skipping")
-                continue
-
-            print("Updating " + data['id'] + " string '" + string_key + "'")
-            data['strings'][string_key][language_to_import] = strings_by_object[data['id']][string_key]
+        previously_translated = translation_existed(target_lang, data['strings'][string_key])
+        translation_needs_update = False
+        if previously_translated:
+            translation_needs_update = translation_changed(verbose, obj_id, string_key,\
+                                          data['strings'][string_key][target_lang],\
+                                          strings_by_object[obj_id][string_key])
+        if not previously_translated or translation_needs_update:
+            print("Updating " + obj_id + " string '" + string_key + "'")
+            data['strings'][string_key][target_lang] = strings_by_object[obj_id][string_key]
             updated = True
```

Comment:
I've changed the behavior here, removing that `TODO` and merging the two branches, I'll paste them here for refrence:
```python
        if not language_to_import in data['strings'][string_key]:
            if strings_by_object[data['id']][string_key] == data['strings'][string_key][fallback_language]:
                # TODO: Is this desirable behaviour?
                if verbose:
                    print("Translation for " + data['id'] + " string '" + string_key + "' is identical to en-GB -- skipping")
                continue

            if verbose:
                print("Adding " + data['id'] + " string '" + string_key + "'")
            data['strings'][string_key][language_to_import] = strings_by_object[data['id']][string_key]
            updated = True
        else:
            if strings_by_object[data['id']][string_key] == data['strings'][string_key][language_to_import]:
                if verbose:
                    print("Translation for " + data['id'] + " string '" + string_key + "' has not changed -- skipping")
                continue

            print("Updating " + data['id'] + " string '" + string_key + "'")
            data['strings'][string_key][language_to_import] = strings_by_object[data['id']][string_key]
```

Reply:
This means we'll have no mechanism avoiding the importing of untranslated (English) strings, correct? I'm a bit torn on this; it seems a bit wasteful. Moreover, if the English string changes in the mean time, the 'translation' won't be automatically updated, right?

Reply:
I see your point, my concern is that it might be annoying for some translators to have the script reporting the translation as incomplete because it was not incorporated due to the fact that it was equal to English (assuming we have everything automatic).

At the same time, if we do the approach I coded, it means that the translations will automatically gain a field duplicating the english text, with no intention.

What if we add something like `"NEEDS_TRANSLATION:"` on the dumping of the strings that fell back to the original language. Then when reading, we can check if it was removed, if it was, it means that the translator acknowledged that this should be kept as English.

Reply:
(By the way, I reverted to the old behavior for now, so we can implement the new one according to our discussion)

Reply:
Per our earlier discussions, I figured it might be something like that, yes. I'm not sure which is the better alternative, to be honest.

Perhaps the script can _leave_ 'English' strings if present, but not _add_ them if somehow identical? That way, we have an override, while the script won't unnecessarily add untranslated strings.

Reply:
> Perhaps the script can leave 'English' strings if present, but not add them if somehow identical? 

Not sure I understood this difference. You mean it won't rewrite English strings that have already been pushed and are equal to the reference, but it won't add any that remains equal?

Reply:
Yes, I think we mean the same thing.  To illustrate, let's say en-GB and nl-NL both use the string "Hamburger" for an object. Dumping nl-NL and reloading it would still leave this string intact. 

Now, let's suppose this string is missing in de-DE. Dumping de-DE would see the reference string from en-GB inserted in its place. However, what I'm proposing is that reloading the new de-DE file would then _not_ create a "Hamburger" string in de-DE.

Reply:
I agree it shouldn't and the current diff doesn't create this new string, so I'm marking this as solved. I'll take this discussion to an issue so we can decide there later and if more people want to chime in :)